# 01 Importing libraries

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02 Importing data

In [4]:
# Strings for path
path=r'C:\Users\Oksana Stepanova\OneDrive\Документи\Instacart Basket Analysis'

In [5]:
# Import data Orders
df_ords=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col=False)

In [6]:
# Import data Products
df_prods=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col=False)

# 03 Data wrangling

In [8]:
# Remove columns from Orders
df_ords.drop(columns=['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [9]:
# Remove columns from Orders and overwrite df_ords
df_ords=df_ords.drop(columns=['eval_set'])

In [10]:
df_ords.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [12]:
# Look for missing values
df_ords['days_since_prior_order'].value_counts(dropna=False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [14]:
# Rename columns in Orders
df_ords.rename(columns={'order_dow':'orders_day_of_week'}, inplace=True)

In [15]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [16]:
# Change data type in Orders
df_ords['order_id']=df_ords['order_id'].astype('str')

In [17]:
df_ords['order_id'].dtype

dtype('O')

# 04 Importing data

In [19]:
# Import data Departments
df_dep=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col=False)

In [20]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [21]:
# Transposing data Departments
df_dep_t=df_dep.T

In [22]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [23]:
# Add index to Departments
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [24]:
# Create new header
new_header=df_dep_t.iloc[0]

In [25]:
new_header

0    department
Name: department_id, dtype: object

In [26]:
# Create a new df and copy everything from the first row
df_dep_t_new=df_dep_t[1:]

In [27]:
# Set new hearder 
df_dep_t_new.columns=new_header

In [28]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [31]:
# Change data type in Orders
df_ords['user_id']=df_ords['user_id'].astype('str')

In [32]:
df_ords['user_id'].dtype

dtype('O')

In [33]:
df_ords.dtypes

order_id                   object
user_id                    object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [35]:
# Count frequency in Orders
df_ords['order_hour_of_day'].value_counts(dropna=False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

# The busies hour to place the order is from 10 am to 11 am

# 05 Data dictionary

In [38]:
# Create Data dictionary of Departments
data_dic_dep=df_dep_t_new.to_dict('index')

In [39]:
# Department name behind department_id '4' 
print(data_dic_dep.get('4'))

{'department': 'produce'}


In [40]:
# Create Data dictionary of Products
data_dict_prods=df_prods.to_dict('index')

In [53]:
print(data_dict_prods.get(['department_id']==4))

{'product_id': 1, 'product_name': 'Chocolate Sandwich Cookies', 'aisle_id': 61, 'department_id': 19, 'prices': 5.8}


# 06 Subsetting

In [55]:
# Breakfast time sales in Orders 
df_ords_bf=df_ords.loc[df_ords['order_hour_of_day'].isin([6,7,8,9])]

In [56]:
df_ords_bf.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count  558410.000000       558410.000000      558410.000000   
mean       18.862005            2.784168           8.187830   
std        18.803509            2.002024           0.899785   
min         1.000000            0.000000           6.000000   
25%         5.000000            1.000000           8.000000   
50%        12.000000            3.000000           8.000000   
75%        26.000000            5.000000           9.000000   
max       100.000000            6.000000           9.000000   

       days_since_prior_order  
count           528379.000000  
mean                10.729049  
std                  8.796813  
min                  0.000000  
25%                  4.000000  
50%                  7.000000  
75%                 14.000000  
max                 30.000000

In [68]:
# Breakfast time sales in Orders on weekend supposing 6 assigned to Saturday and 0 assignd to Sunday 
df_ords_bf_wend=df_ords.loc[(df_ords['orders_day_of_week'].isin([0,6]))&(df_ords['order_hour_of_day'].isin([6,7,8,9]))]

In [69]:
df_ords_bf_wend.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count  153006.000000       153006.000000      153006.000000   
mean       17.700378            2.680718           8.227207   
std        17.662194            2.982971           0.862250   
min         1.000000            0.000000           6.000000   
25%         5.000000            0.000000           8.000000   
50%        12.000000            0.000000           8.000000   
75%        25.000000            6.000000           9.000000   
max       100.000000            6.000000           9.000000   

       days_since_prior_order  
count           144249.000000  
mean                11.284397  
std                  8.827176  
min                  0.000000  
25%                  5.000000  
50%                  8.000000  
75%                 15.000000  
max                 30.000000

In [70]:
# Breakfast time sales in Orders on workdays supposing 1,2,3,4,5 assigned to workdays from Monday to Friday 
df_ords_bf_workd=df_ords.loc[(df_ords['orders_day_of_week'].isin([1,2,3,4,5]))&(df_ords['order_hour_of_day'].isin([6,7,8,9]))]

In [71]:
df_ords_bf_workd.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count  405404.000000       405404.000000      405404.000000   
mean       19.300421            2.823211           8.172968   
std        19.198408            1.468666           0.913110   
min         1.000000            1.000000           6.000000   
25%         5.000000            1.000000           8.000000   
50%        13.000000            3.000000           8.000000   
75%        27.000000            4.000000           9.000000   
max       100.000000            5.000000           9.000000   

       days_since_prior_order  
count           384130.000000  
mean                10.520503  
std                  8.776325  
min                  0.000000  
25%                  4.000000  
50%                  7.000000  
75%                 14.000000  
max                 30.000000

In [67]:
# Count of values in 'day of week' column
df_ords['orders_day_of_week'].value_counts(dropna=False)

orders_day_of_week
0    600905
1    587478
2    467260
5    453368
6    448761
3    436972
4    426339
Name: count, dtype: int64

# The subsets above show the number of orders placed during breakfast, presumably between 6am and 9am during the week and separately on weekends and weekdays. The number of orders on Saturday and Sunday is 27% of the total number of orders. I expected to see most orders placed on weekends. However, the frequency is likely distributed evenly throughout the week.

In [81]:
# Department_id related to alcohol, deli, beverages, meat/seafood
df_dep_t_new.loc[df_dep_t_new['department'].isin(['alcohol', 'deli', 'beverages', 'meat seafood'])] 

department_id    department
5                   alcohol
7                 beverages
12             meat seafood
20                     deli

In [82]:
# Subset of products available in alcohol, deli, beverages, meat/seafood departments
df_prods_parties=df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [84]:
df_prods_parties

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

# The dataframe above shows the products that customers might use for dinner parties. It represents 7650 rows and 5 columns 

In [89]:
# Select all rows from Orders where user_id=1
df_ords_user_id_1=df_ords.loc[df_ords['user_id']=='1']

In [90]:
df_ords_user_id_1

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

In [93]:
df_ords_user_id_1.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

In [106]:
# Count frequency of orders by day of week 
df_ords_user_id_1['orders_day_of_week'].value_counts(dropna=False)

orders_day_of_week
4    4
1    3
2    2
3    2
Name: count, dtype: int64

In [107]:
# Count frequency of orders by hour of day 
df_ords_user_id_1['order_hour_of_day'].value_counts(dropna=False)

order_hour_of_day
8     3
7     3
12    1
15    1
9     1
14    1
16    1
Name: count, dtype: int64

# The customer with user_id=1 (i) makes orders between 1st and 4th days of week and never places the orders other days; (ii) places the order between 7 am and 4 pm; (iii) the most often orders on the 4th day of the week between 7 am and 8 am;  (iv) on average, orders products every 19 days, the longest period of time when client hasn’t made any purchases was 30 days.

# 07 Exporting dataframes

In [113]:
# Export Orders dataframe
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

In [115]:
# Export Departments dataframe
df_dep_t_new.to_csv(os.path.join(path,'02 Data', 'Prepared Data', 'departments_wrangled.csv'))